docs/dev/guidelines/server.rst

```diff
@@ -350,3 +350,101 @@ Please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-c
+* Write self-contained tests by revealing all relevant parameters, insert data right in the test and prefer composition over inheritance.
+* Write dumb tests by avoiding the reuse of production code and focusing on comparing output values with hard-coded values.
+* KISS > DRY ("Keep it simple, Stupid!" and "Don't repeat yourself!")
+* Test close to production by focusing on testing a complete vertical slide and avoiding in-memory databases.
+* JUnit5 and AssertJ are a very good choice.
+* Invest in a testable implementation by avoiding static access, using constructor injection, using Clocks and separating business logic from asynchronous execution.
+
+For a more detail overview definitely check out: https://phauer.com/2019/modern-best-practices-testing-java/
+
+
+Make use of JUnit 5 Features:
+https://junit.org/junit5/docs/current/user-guide/#writing-tests
+https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html
+
+Here you can find JUnit 5 best practices:
+https://howtodoinjava.com/best-practices/unit-testing-best-practices-junit-reference-guide/
+
+Also check out this page for spring related testing:
+https://www.baeldung.com/spring-tests
+
+If you want to write tests for Programming Exercises to test student's submission check out `this <https://confluence.ase.in.tum.de/display/ArTEMiS/Best+Practices+for+writing+Java+Programming+Exercise+Tests+in+Artemis>`__.
+
+
+23. Avoid using @MockBean
+=========================
+
+Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here <https://www.baeldung.com/spring-tests>`__.
+Basically every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
+This leads to a large overhead, which makes the tests tend to take a lot more time.
+
+Here is an example how to replace a ``@SpyBean``. We wanted to test a catch case which is only executed if an IOException gets thrown. We did this by mocking the service method and making it throw an Exception.
+
+.. code-block:: java
+
+    public class TestExport extends AbstractSpringIntegrationBambooBitbucketJiraTest {
+        @SpyBean
+        private FileUploadSubmissionExportService fileUploadSubmissionExportService;
+
+        @Test
+        @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+        public void testExportAll_IOException() throws Exception {
+            doThrow(IOException.class).when(fileUploadSubmissionExportService).export(any(), any());
+            request.postWithResponseBodyFile("/api/file-upload-export/" + fileUploadExercise.getId()", HttpStatus.BAD_REQUEST);
+        }
+    }
+
+This, as mentioned above, we should really avoid.
```

Comment:
As mentioned above, we should really avoid this

Reply:
I changed it :+1: